File: ruby/sieve/sieve.rb
# frozen_string_literal: true

#
# Implementation of the Sieve of Eratosthenes
#
class Sieve
  def initialize(limit)
    raise ArgumentError, "#{limit} is not a valid limit." if limit < 1

    @limit = limit
  end

  def primes
    primes = (2..@limit).to_a
    n = primes.length

    (0...n).each do |i|
      base = primes[i]
      next if base.nil?

      ((i + base)...n).step(base) { |j| primes[j] = nil }
    end

    primes.compact
  end

  def self.primes(limit)
    new(limit).primes
  end
end

File: java/sieve/Sieve.java
import java.util.ArrayList;
import java.util.Objects;
import java.util.List;
import java.util.stream.IntStream;

class Sieve {
    int limit;
    Sieve(int maxPrime) {
        if (maxPrime < 1) {
            throw new IllegalArgumentException(String.format(
                "%d is not a valid limit.", maxPrime));
        }
        limit = maxPrime;
    }

    List<Integer> getPrimes() {
        ArrayList<Integer> primes = new ArrayList<>();
        // populate primes array
        IntStream.rangeClosed(2, limit).forEach(primes::add);

        for (int i = 0, n = primes.size(); i < n; i++) {
            Integer base = primes.get(i);
            if (base == null) { // number is composite
                continue;
            }

            // relies that the interval space is 1. ie, the space between two
            // numbers and their indices is the same.
            for (int j = i + base; j < n; j += base) {
                // number is a multiple of base, and thus, composite
                primes.set(j, null);
            }
        }

        primes.removeIf(Objects::isNull); // remove composite placeholders

        return primes;
    }

    // the way it should be in a sane programming language. just one function
    // with one parameter
    static List<Integer> getPrimes(int limit) {
        return new Sieve(limit).getPrimes();
    }
}
